compiler/src/iree/compiler/Codegen/Common/test/canonicalize_early_bufferization_ops.mlir

```diff
@@ -0,0 +1,178 @@
+//   CHECK-DAG:   %[[COLLAPSE:.+]] = memref.collapse_shape %[[SRC_SUBSPAN]]
+//   CHECK-DAG:   %[[EXPAND:.+]] = memref.expand_shape %[[COLLAPSE]]
+//  CHECK-SAME:     output_shape [%[[D0]], %[[D1]]]
+//   CHECK-DAG:   %[[LOAD:.+]] = iree_codegen.load_from_memref %[[EXPAND]]
+//   CHECK-DAG:   %[[DEST_SUBSPAN:.+]] = hal.interface.binding.subspan{{.*}} binding(1)
+//   CHECK-DAG:   %[[BARRIER:.+]] = util.optimization_barrier %[[LOAD]]
+//       CHECK:   iree_codegen.store_to_memref %[[BARRIER]], %[[DEST_SUBSPAN]]
+
+// -----
+
+#pipeline_layout = #hal.pipeline.layout<constants = 4, bindings = [
+  #hal.pipeline.binding<storage_buffer>,
+  #hal.pipeline.binding<storage_buffer>
+]>
+func.func @fold_dynamic_reshape_store() {
+  %c0 = arith.constant 0 : index
+  %0 = hal.interface.constant.load layout(#pipeline_layout) ordinal(0) : index
+  %1 = hal.interface.constant.load layout(#pipeline_layout) ordinal(1) : index
+  %2 = hal.interface.constant.load layout(#pipeline_layout) ordinal(2) : index
+  %3 = hal.interface.constant.load layout(#pipeline_layout) ordinal(3) : index
+  %4 = hal.interface.binding.subspan layout(#pipeline_layout) binding(0) : memref<?x?xf32, #hal.descriptor_type<storage_buffer>>{%0, %1}
+  memref.assume_alignment %4, 1 : memref<?x?xf32, #hal.descriptor_type<storage_buffer>>
+  %5 = hal.interface.binding.subspan layout(#pipeline_layout) binding(1) : memref<?x?xf32, #hal.descriptor_type<storage_buffer>>{%2, %3}
+  memref.assume_alignment %5, 1 : memref<?x?xf32, #hal.descriptor_type<storage_buffer>>
+  %6 = iree_codegen.load_from_memref %4 : memref<?x?xf32, #hal.descriptor_type<storage_buffer>> -> tensor<?x?xf32>
+  %barrier = util.optimization_barrier %6 : tensor<?x?xf32>
+  %collapsed = tensor.collapse_shape %barrier [[0, 1]] : tensor<?x?xf32> into tensor<?xf32>
+  %expanded = tensor.expand_shape %collapsed [[0, 1]] output_shape [%2, %3] : tensor<?xf32> into tensor<?x?xf32>
+  iree_codegen.store_to_memref %expanded, %5 : tensor<?x?xf32> into memref<?x?xf32, #hal.descriptor_type<storage_buffer>>
+  return
+}
+// CHECK-LABEL: @fold_dynamic_reshape_store
+//   CHECK-DAG:   %[[C0:.+]] = arith.constant 0
+//   CHECK-DAG:   %[[C1:.+]] = arith.constant 1
+//   CHECK-DAG:   %[[SRC_SUBSPAN:.+]] = hal.interface.binding.subspan{{.*}} binding(0)
+//   CHECK-DAG:   %[[LOAD:.+]] = iree_codegen.load_from_memref %[[SRC_SUBSPAN]]
+//   CHECK-DAG:   %[[BARRIER:.+]] = util.optimization_barrier %[[LOAD]]
+//   CHECK-DAG:   %[[D0:.+]] = tensor.dim %[[BARRIER]], %[[C0]]
+//   CHECK-DAG:   %[[D1:.+]] = tensor.dim %[[BARRIER]], %[[C1]]
+//   CHECK-DAG:   %[[DEST_SUBSPAN:.+]] = hal.interface.binding.subspan{{.*}} binding(1)
+//   CHECK-DAG:   %[[COLLAPSE:.+]] = memref.collapse_shape %[[DEST_SUBSPAN]]
+//   CHECK-DAG:   %[[EXPAND:.+]] = memref.expand_shape %[[COLLAPSE]]
+//  CHECK-SAME:     output_shape [%[[D0]], %[[D1]]]
+//       CHECK:   iree_codegen.store_to_memref %[[BARRIER]], %[[EXPAND]]
```

Comment:
maybe add the types here as well so its easier for the reader to see the resulting IR?